packages/git/src/browser/git-repository-provider.ts

```diff
@@ -43,8 +47,13 @@ export class GitRepositoryProvider {
         this.workspaceService.onWorkspaceChanged(event => {
             this.refresh();
         });
-        await this.refresh({ maxCount: 1 });
+        this._selectedRepository = await this.storageService.getData<Repository | undefined>(this.selectedRepoStorageKey);
+        this._allRepositories = await this.storageService.getData<Repository[]>(this.allRepoStorageKey);
+        if (!this._allRepositories) {
+            await this.refresh({ maxCount: 1 });
+        }
         await this.refresh();
+        this.storageService.setData<Repository[]>(this.allRepoStorageKey, this._allRepositories);
```

Comment:
We should we store allRepositories? They should be read from disk, no?
Also storing them on initialize doesn't seem right.

Reply:
We want to use them to quickly render a list of the repositories instead of using `this.refresh({ maxCount: 1 });`. 
As @akosyakov stated [here](https://github.com/theia-ide/theia/pull/3007#discussion_r221420403) that was needed especially for big projects with nested repositories. But the `{maxCount: 1}` option prevents restoring a selected repository (unless it is a root repository). 
So, we need to store allRepositories as clients could wait too long in such big projects.

Regarding setting them while initialisation: ok, I guess it should be ok to set them in refresh method.